pkg/kubectl/cmd/set/set_selector.go

```diff
@@ -0,0 +1,280 @@
+		return err
+	}
+
+	builder := resource.NewBuilder(o.Mapper, o.Typer, resource.ClientMapperFunc(f.ClientForMapping), f.Decoder(true)).
+		ContinueOnError().
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(enforceNamespace, o.Recursive, o.Filenames...).
+		Flatten()
+	if !o.Local {
+		builder = builder.
+			ResourceTypeOrNameArgs(o.All, o.Resources...).
+			Latest()
+	}
+	o.Infos, err = builder.Do().Infos()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Validate basic inputs
+func (o *SelectorOptions) Validate() error {
+	if len(o.Resources) < 1 && len(o.Filenames) == 0 {
+		return fmt.Errorf("one or more resources must be specified as <resource> <name> or <resource>/<name>")
+	}
+	return nil
+}
+
+// copyMatchLabels copy match labels from src to dst
+func copyMatchLabels(dst *map[string]string, src *map[string]string, overwrite bool) error {
+	if overwrite {
+		*dst = make(map[string]string)
+	}
+	if !overwrite {
+		for label := range *src {
+			_, exists := (*dst)[label]
+			if exists {
+				return fmt.Errorf("MatchLabel %v exists, use --overwrite to replace it", label)
+			}
+		}
+	}
+	for label, value := range *src {
+		_, exists := (*dst)[label]
+		if overwrite || !exists {
+			(*dst)[label] = value
+		}
+	}
+	return nil
+}
+
+// copySelector copy match labels and match expressions from src to dst
+func copySelector(dst *unversioned.LabelSelector, src *unversioned.LabelSelector, overwrite bool) error {
+	copyMatchLabels(&dst.MatchLabels, &src.MatchLabels, overwrite)
+	if overwrite || dst.MatchExpressions == nil {
+		dst.MatchExpressions = make([]unversioned.LabelSelectorRequirement, len(src.MatchExpressions))
+	}
+	if overwrite {
+		copy(dst.MatchExpressions, src.MatchExpressions)
+		return nil
+	}
+	for _, srcExp := range src.MatchExpressions {
+		for _, dstExp := range dst.MatchExpressions {
+			if srcExp.Key == dstExp.Key {
+				return fmt.Errorf("MatchExpression %v exists, use --overwrite to replace it", srcExp.Key)
+			}
+		}
+	}
+	for _, srcExp := range src.MatchExpressions {
+		dst.MatchExpressions = append(dst.MatchExpressions, srcExp)
+	}
+	return nil
+}
+
+// Run the command.
+func (o *SelectorOptions) Run() error {
+	allErrs := []error{}
+
+	patches := CalculatePatches(o.Infos, o.Encoder, func(info *resource.Info) (bool, error) {
+		err := o.UpdateSelectorForObject(info.Object, func(map_selector *map[string]string, selector *unversioned.LabelSelector) error {
+			if map_selector != nil {
+				if len(o.Selector.MatchExpressions) > 0 {
+					return fmt.Errorf("error: MatchExpression %v not supported on this object", o.Selector.MatchExpressions)
+				}
+				err := copyMatchLabels(map_selector, &o.Selector.MatchLabels, o.Overwrite)
+				if err != nil {
+					return err
+				}
+			}
+			if selector != nil {
+				err := copySelector(selector, o.Selector, o.Overwrite)
+				if err != nil {
+					return err
+				}
+			}
+
+			return nil
+		})
+		return true, err
+	})
+
+	for _, patch := range patches {
```

Comment:
I think I'd like to see this method use `.Visit` like most of the other CLI commands that deal with resources from a builder.